content/rationale.md

```diff
@@ -77,6 +77,15 @@ the module name and retains function names in the case format they are declared
 in. Parameters names are inconsistently mapped, preferring the case format of
 the language.
 
+## Why are the query parameters URI encoded?
```

Comment:
How about this @codefromthecrypt ?

Reply:
looks good. only thing left is clarifying why we do this host side.

Ex. it isn't really the user who is calling set_uri, it is the guest code. so, the guest SDKs are allowed to not encode the inputs and let the host do it instead. The host MUST ensure a raw query is encoded to prevent the issue you are mentioning. The reason why we allow the guest to opt-out is to keep their code size down. If we instead said the guest must encode, their wasm would bloat with logic to perform encoding, possibly even allocating, resulting in a more complex minimal guest.

TL;DR; this is right about why we need to encode, and we can also add that push this to the host to make guest SDKs easier and leaner. e.g. without the host doing this, it would be near impractical to implement anything in `%.wat`, and fuller featured languages like tinygo would both have more code and higher overhead to attempt to encode first.

Reply:
AFAIK, both get_uri and set_uri have the same ambiguity issue, `set_uri("/disney?name=chip&dale")`, with the intention of sending two query parameters, `name` and `dale`, if the host encodes this it will become `set_uri("/disney?name=chip%26dale")` with just one parameter `name`.

I don't see a way around this without expecting the guest to pass an encoded URL. That doesn't mean the guest SDK has to encode, hence https://github.com/http-wasm/http-wasm-guest-tinygo/pull/22, and the user could use hard-coded encoded strings to avoid code or for use in `%.wat` if that's viable. Logic acting on non-hardcoded strings will likely need to bring in the code though.

On the bright side, none of our current hosts encode a passed in URI from what I can tell, for example we parse an encoded URI here

https://github.com/http-wasm/http-wasm-host-go/blob/main/handler/nethttp/host.go#L67